web/src/app/rotations/components/SetFavoriteButton.js

```diff
@@ -0,0 +1,109 @@
+import React from 'react'
+import p from 'prop-types'
+import gql from 'graphql-tag'
+import { Mutation } from 'react-apollo'
+import IconButton from '@material-ui/core/IconButton'
+import FavoriteFilledIcon from '@material-ui/icons/Star'
+import FavoriteBorderIcon from '@material-ui/icons/StarBorder'
+import { graphql2Client } from '../../apollo'
+import Query from '../../util/Query'
+
+const query = gql`
+  query favQuery($id: ID!) {
+    rotation(id: $id) {
+      id
+      isFavorite
+    }
+  }
+`
+
+const mutation = gql`
+  mutation setFav($input: SetFavoriteInput!) {
+    setFavorite(input: $input)
+  }
+`
+
+export default class SetFavoriteButton extends React.PureComponent {
```

Comment:
I would recommend that we, instead of having multiple components named `SetFavoriteButton` under multiple directories within GoAlert with the majority of the same code between them, we consolidate these files into one (probably under `app/util`).

It could accept props that are more generic, such as `targetID: p.string.isRequired` as well as `targetType: p.oneOf(['service, rotation'])` which can be used to determine the data key, as well as updating the cache.

This file can also be written as a React hook, now that we're on the latest React version.

Reply:
What would be the best practice for the queries and mutations when moving this to util? Should we ask for a query and mutation prop?

Reply:
Yeah, that's what we do in other locations. For example we have a generic `FormDialog` component, but we send it the form as a prop